# Data Integration and Cleaning with SQLAlchemy and Pandas

## Introduction

This guide demonstrates how to integrate data from a MySQL database into a Python environment using SQLAlchemy and Pandas. It outlines the installation of required Python libraries, connection to the database, data loading, cleaning, and saving the cleaned dataset back into the database.

The steps include:
1. Installing the necessary packages.
2. Connecting to a MySQL database.
3. Loading data into a Pandas DataFrame.
4. Cleaning and transforming the data.
5. Saving the cleaned data back to the database.

## Execution Output

### Step 1: Install Required Libraries

```bash
pip install SQLAlchemy
pip install pymysql


In [2]:
pip install SQLAlchemy

   ---------------------------------------- 0.0/2.1 MB ? eta -:--:--
   --------------- ------------------------ 0.8/2.1 MB 4.8 MB/s eta 0:00:01
   ------------------------------ --------- 1.6/2.1 MB 4.2 MB/s eta 0:00:01
   ---------------------------------------- 2.1/2.1 MB 4.3 MB/s eta 0:00:00
Note: you may need to restart the kernel to use updated packages.


In [1]:
pip install pymysql

In [2]:
import pandas as pd
from sqlalchemy import create_engine

In [7]:
db_connection_str = 'mysql+pymysql://root:password@localhost:3306/CustomerData'

In [8]:
# Create the connection
db_connection = create_engine(db_connection_str)

In [9]:
# Load data from MySQL into a Pandas DataFrame
df = pd.read_sql('SELECT * FROM StagingTable', con=db_connection)

In [10]:
df.head()

CustomerID Age  Gender Tenure UsageFrequency SupportCalls PaymentDelay  \
0          2  30  Female     39             14            5           18   
1          3  65  Female     49              1           10            8   
2          4  55  Female     14              4            6           18   
3          5  58    Male     38             21            7            7   
4          6  23    Male     32             20            5            8   

  SubscriptionType ContractLength TotalSpend LastInteraction Churn  
0         Standard         Annual        932              17   1\r  
1            Basic        Monthly        557               6   1\r  
2            Basic      Quarterly        185               3   1\r  
3         Standard        Monthly        396              29   1\r  
4            Basic        Monthly        617              20   1\r

In [11]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 440833 entries, 0 to 440832
Data columns (total 12 columns):
 #   Column            Non-Null Count   Dtype 
---  ------            --------------   ----- 
 0   CustomerID        440832 non-null  object
 1   Age               440832 non-null  object
 2   Gender            440832 non-null  object
 3   Tenure            440832 non-null  object
 4   UsageFrequency    440832 non-null  object
 5   SupportCalls      440832 non-null  object
 6   PaymentDelay      440832 non-null  object
 7   SubscriptionType  440832 non-null  object
 8   ContractLength    440832 non-null  object
 9   TotalSpend        440832 non-null  object
 10  LastInteraction   440832 non-null  object
 11  Churn             440833 non-null  object
dtypes: object(12)
memory usage: 40.4+ MB


In [12]:
# Check unique values in each column
for column in df.columns:
    print(f"Unique values in {column}:")
    print(df[column].unique())
    print("\n")

Unique values in CustomerID:
['2' '3' '4' ... '449997' '449998' '449999']


Unique values in Age:
['30' '65' '55' '58' '23' '51' '39' '64' '29' '52' '22' '48' '24' '49'
 '19' '47' '42' '57' '27' '59' '21' '60' '35' '18' '56' '20' '63' '25'
 '28' '32' '38' '37' '31' '53' '41' '33' '26' '36' '44' '34' '61' '40'
 '45' '46' '54' '43' '50' '62' None]


Unique values in Gender:
['Female' 'Male' None]


Unique values in Tenure:
['39' '49' '14' '38' '32' '33' '37' '12' '3' '18' '21' '41' '35' '4' '56'
 '44' '15' '55' '43' '52' '26' '2' '29' '59' '40' '51' '53' '24' '30' '6'
 '28' '17' '60' '7' '34' '10' '5' '45' '54' '58' '25' '13' '47' '31' '22'
 '19' '23' '1' '8' '46' '16' '50' '48' '11' '42' '27' '9' '20' '57' '36'
 None]


Unique values in UsageFrequency:
['14' '1' '4' '21' '20' '25' '12' '8' '5' '9' '6' '17' '23' '13' '16' '27'
 '2' '28' '29' '15' '24' '3' '22' '26' '30' '7' '11' '18' '19' '10' None]


Unique values in SupportCalls:
['5' '10' '6' '7' '9' '3' '4' '2' '0' '1' '8' None]


Un

In [13]:
# Replace '1\r' with '1' and '0\r' with '0' in the 'Churn' column
df['Churn'] = df['Churn'].str.replace('1\r', '1').str.replace('0\r', '0')


In [14]:
# Check unique values in each column
for column in df.columns:
    print(f"Unique values in {column}:")
    print(df[column].unique())
    print("\n")

Unique values in CustomerID:
['2' '3' '4' ... '449997' '449998' '449999']


Unique values in Age:
['30' '65' '55' '58' '23' '51' '39' '64' '29' '52' '22' '48' '24' '49'
 '19' '47' '42' '57' '27' '59' '21' '60' '35' '18' '56' '20' '63' '25'
 '28' '32' '38' '37' '31' '53' '41' '33' '26' '36' '44' '34' '61' '40'
 '45' '46' '54' '43' '50' '62' None]


Unique values in Gender:
['Female' 'Male' None]


Unique values in Tenure:
['39' '49' '14' '38' '32' '33' '37' '12' '3' '18' '21' '41' '35' '4' '56'
 '44' '15' '55' '43' '52' '26' '2' '29' '59' '40' '51' '53' '24' '30' '6'
 '28' '17' '60' '7' '34' '10' '5' '45' '54' '58' '25' '13' '47' '31' '22'
 '19' '23' '1' '8' '46' '16' '50' '48' '11' '42' '27' '9' '20' '57' '36'
 None]


Unique values in UsageFrequency:
['14' '1' '4' '21' '20' '25' '12' '8' '5' '9' '6' '17' '23' '13' '16' '27'
 '2' '28' '29' '15' '24' '3' '22' '26' '30' '7' '11' '18' '19' '10' None]


Unique values in SupportCalls:
['5' '10' '6' '7' '9' '3' '4' '2' '0' '1' '8' None]


Un

In [15]:
# Convert numeric columns to their appropriate types, replacing invalid values with NaN
df['Age'] = pd.to_numeric(df['Age'], errors='coerce')
df['Tenure'] = pd.to_numeric(df['Tenure'], errors='coerce')
df['UsageFrequency'] = pd.to_numeric(df['UsageFrequency'], errors='coerce')
df['SupportCalls'] = pd.to_numeric(df['SupportCalls'], errors='coerce')
df['PaymentDelay'] = pd.to_numeric(df['PaymentDelay'], errors='coerce')
df['TotalSpend'] = pd.to_numeric(df['TotalSpend'], errors='coerce')
df['LastInteraction'] = pd.to_numeric(df['LastInteraction'], errors='coerce')

In [16]:
df.isna().sum()

CustomerID          1
Age                 1
Gender              1
Tenure              1
UsageFrequency      1
SupportCalls        1
PaymentDelay        1
SubscriptionType    1
ContractLength      1
TotalSpend          1
LastInteraction     1
Churn               0
dtype: int64

In [18]:
# Save the cleaned data back into the MySQL database, using the correct table name
df.to_sql('stagingtable', con=db_connection, if_exists='replace', index=False)

440833

## Output Summary

- **SQLAlchemy and PyMySQL Installed Successfully**: Required packages (`SQLAlchemy`, `pymysql`) were installed.
- **Database Connection Established**: Connected to MySQL database `CustomerData` using SQLAlchemy.
- **Data Loaded**: Successfully loaded data from `StagingTable` into a Pandas DataFrame with 440,833 rows and 12 columns.
- **Data Cleaned**:
  - Fixed trailing characters in the `Churn` column.
  - Converted numeric columns to appropriate types.
  - Identified and handled missing values.
- **Data Saved**: Cleaned data written back to `stagingtable` in the MySQL database.
- **Final Record Count**: 440,833 rows saved.
